src/request-queue.ts

```diff
@@ -1,99 +1,90 @@
-             *  and it is empty. To avoid empty polling (user went idle, stepped away from comp)
-             *  we can turn it off with the isPolling flag.
-             *
-             * Polling will be re enabled when the next time PostHogLib.capture is called with
-             *  an event that should be added to the event queue.
-             */
-            if (this._empty_queue_count > 4) {
-                this.isPolling = false
-                this._empty_queue_count = 0
-            }
-            if (this.isPolling) {
-                this.poll()
-            }
-        }, this._pollInterval) as any as number
-    }
-
     unload(): void {
-        clearTimeout(this._poller)
-        const requests = this._event_queue.length > 0 ? this.formatQueue() : {}
-        this._event_queue.length = 0
+        this.clearFlushTimeout()
+        const requests = this.queue.length > 0 ? this.formatQueue() : {}
+        this.queue.length = 0
         const requestValues = Object.values(requests)
 
         // Always force events to be sent before recordings, as events are more important, and recordings are bigger and thus less likely to arrive
         const sortedRequests = [
             ...requestValues.filter((r) => r.url.indexOf('/e') === 0),
             ...requestValues.filter((r) => r.url.indexOf('/e') !== 0),
         ]
-        sortedRequests.map(({ url, data, options }) => {
-            this.handlePollRequest(url, data, { ...options, transport: 'sendBeacon' })
+        sortedRequests.map((req) => {
+            this.sendRequest({ ...req, transport: 'sendBeacon' })
         })
     }
 
-    formatQueue(): Record<string, QueuedRequestData> {
-        const requests: Record<string, QueuedRequestData> = {}
-        _each(this._event_queue, (request) => {
-            const { url, data, options } = request
-            const key = (options ? options._batchKey : null) || url
-            if (_isUndefined(requests[key])) {
-                requests[key] = { data: [], url, options }
+    enable(): void {
+        this.isPaused = false
+        this.setFlushTimeout()
+    }
+
+    private setFlushTimeout(): void {
+        if (this.isPaused) {
+            return
+        }
+        this.flushTimeout = setTimeout(() => {
+            this.clearFlushTimeout()
+            if (this.queue.length > 0) {
+                const requests = this.formatQueue()
+                for (const key in requests) {
+                    const req = requests[key]
+                    const now = new Date().getTime()
+
+                    if (req.data && _isArray(req.data)) {
+                        _each(req.data, (data) => {
+                            data['offset'] = Math.abs(data['timestamp'] - now)
+                            delete data['timestamp']
+                        })
+                    }
+                    this.sendRequest(req)
+                }
+                this.queue = []
```

Comment:
so if I'm reading it right `formatQueue` is copying the queue, then we do each request, and then we clear the queue...

should we clear the queue first? is there a risk that `sendRequest` has enqueued something that we then clear?

Reply:
Should be impossible for sendRequest to add more to the queue but nonetheless I'll move it around to be more sure of it